syndeo-template/src/main/java/com/google/cloud/syndeo/transforms/files/SyndeoFilesReadSchemaTransformProvider.java

```diff
@@ -0,0 +1,287 @@
+            fileDeserializer(configuration.getFormat(), configuration.getSchema());
+        return new PTransform<PCollectionRowTuple, PCollectionRowTuple>() {
+          @Override
+          public PCollectionRowTuple expand(PCollectionRowTuple input) {
+            PCollectionTuple firstResult =
+                input
+                    .getPipeline()
+                    .apply(
+                        PubsubIO.readStrings()
+                            .fromSubscription(configuration.getPubsubSubscription()))
+                    .apply(
+                        ParDo.of(
+                                new DoFn<String, KV<String, Long>>() {
+                                  @ProcessElement
+                                  public void process(
+                                      @Element String fileNotification,
+                                      MultiOutputReceiver receiver) {
+                                    ObjectMapper om = new ObjectMapper();
+                                    try {
+                                      JsonNode notificationPayload = om.readTree(fileNotification);
+                                      String fileName =
+                                          "gs://"
+                                              + Path.of(
+                                                  notificationPayload.get("bucket").asText(),
+                                                  notificationPayload.get("name").asText());
+                                      receiver
+                                          .get(FILE_GENERATION_OUTPUT_TAG)
+                                          .output(
+                                              KV.of(
+                                                  fileName,
+                                                  notificationPayload
+                                                      .get("generation")
+                                                      .longValue()));
+                                    } catch (JsonProcessingException e) {
+                                      receiver
+                                          .get(ERROR_TAG)
+                                          .output(
+                                              Row.withSchema(ERROR_SCHEMA)
+                                                  .addValues(fileNotification, e.toString())
+                                                  .build());
+                                    }
+                                  }
+                                })
+                            .withOutputTags(
+                                FILE_GENERATION_OUTPUT_TAG, TupleTagList.of(ERROR_TAG)));
+
+            PCollectionTuple result =
+                firstResult
+                    .get(FILE_GENERATION_OUTPUT_TAG)
+                    .apply(
+                        Deduplicate.<KV<String, Long>>values()
+                            .withDuration(Duration.standardMinutes(5L)))
```

Comment:
is this duration enough for dedupe? what happens if the notification is delayed, would we be consuming the file again?
Also, what would be window interval for the pipeline?

Reply:
the pipeline itself would all be on the GlobalWindow - this Duration would be a 'real time' duration of 5 minutes. If a duplicate of a notification arrives over 5 minutes after the initial message, then the file would indeed be consumed twice. We can expand this to be a larger number. I don't think a larger time would incur a significant larger cost.

Reply:
i see, thanks for the explanation. Would we need to change the window duration and the dedupe both in that case to a larger number? Just changing the dedupe interval without the windowing duration would not be enough, right?